### ISSUE TITLE
Number of labels for ranking.

### ISSUE THREAD

**Issue by nboissin** [microsoft/LightGBM#4078]
I am using LGBMRanker on a ranking problem that requires more than 30 labels, lets say 600.
i would like to avoid making 30 classes of 20 labels because it downgrades the results.
I have set label_gain=2**np.arange(0,600)
Exist ndgc function returns 1.
I have provided my own ndgc calculation in case existing one would have an overflow issue with 2^31, but the problem is still there. The origin of the problem is that all predicted values are 0.
is there another parameter that i should set?
The issue is not new, but i have not found any answer to my problem.

**Comment by jameslamb** [comment]
Hi @nboissin , thanks for using LightGBM! Could you please provide a minimal reproducible example, some code that we could run to understand what you're trying to do? It would help us to give you a clear answer, thanks.

**Comment by shiyu1994** [comment]
@nboissin Thanks for using LightGBM. Although `2^600` is within the range of 64-bit double, LightGBM uses 32-bit single precision floating point numbers for gradients and hessians.

So I think if you set the `label_gain` by `2 ** np.arange(0, 600)`, there'll be a lot of `inf` or `-inf` in the gradients and hessians during boosting.

There're several solutions for large number of ranking labels, according to #2892.

If you have any further problem, please feel free to post it here.

**Comment by nboissin** [comment]
Thank you for your answer. +-inf hessians is probably the source of my problem.
I went through  #2892, but there is no solution proposed there (it only adresses the label_gain setting). Would you have any other pointers. My problem is by structure a ranking problem, and LightGBM is already providing great results when I group my labels in 30 groups, I would really like to test it with more accurates ranks. Any pointers that i could read?

**Comment by rripken** [comment]
I'm encountering a similar issue and could use some suggestions.  I have about 300 results for each of my queries and each result has a floating point fitness measure.
I can use pandas to group by the query id and then put the results into order, something like:
```
df['place'] = df.groupby('query_id')['fitness'].transform(lambda x: pd.factorize(x, sort=True)[0]+1)
df['label'] = 300 - df['place']
```
After this I have a label column with positive integer values where better results have higher values.

But I get an error message "Label 285 is not less than the number of label mappings (31)".   If I understand #2892 and related issues correctly the problem is that the default label_gain array has 31 entries in it and lightgbm has detected that I'm trying to use more than 31 different labels.  In my case its complaining about 285 because that is the first value lightgbm saw that was larger than the default label_gain array's size.  Is that an accurate interpretation of what the error I'm seeing means?

In #2892 @guolinke wrote "in ranking problems, we usually focus on the top docs" which is starting to make some ideas click for me.  Does the default label_gain built into lightgbm only include 31 entries because by the time the ranker gets to results in 30th and 31st place the ranker is paying so little attention to them that they don't really matter? 
If that is the case what can/should be done? 
I could just drop all the items with a "place" > 31 - but I'd rather not do that because I'd be throwing out about 90% of my dataset and I think there is useful information to be learned from those results.

The discussion in #2892 suggests relabeling the results with 1-4.  I don't understand why 1-4 would be used and not 1-30 but it appears @nboissin used 30 groups and got great results so perhaps the 1-4 example was just a poor example.  When binning would it make sense to use equal size bins?  Or would I want to "focus on the top docs" and use more but smaller bins for the most relevant results and then have increasing larger bins as I progressed, ending with bin 0 having something like the worst 25%?  

There is additional discussion of providing a different label_gain array.  Is there documentation anywhere that describes how label_gain should function?   I see that the default array starts [0,1,3,7...]. The values look like (2^^x -1).  In other issues I've seen that the values should fit in a 32 bit int.  Those two things explain why there are only 31 values in the default label_gain.  Is there a reason that the weights exponentially increase?  Is that where the "focus on the top docs" comes from?  I've seen some mention people also trying a linear gain but no real discussion of when or why that should be done.  If I wanted to have a similar sort of exponential ramp up but 300 items in the label_gain I assume I could just use a different base?  I assume the exponential nature is most important at the end of the label_gain - where the gains for the most relevant weights are located?  I'm also assuming that the worst result (label_gain[0]) should be 0 and that the most relevant result can be as large as 2**31 -1 ?

Something like:

```
def get_label_gain(count):
    retval = []
    retval.append(0)
    retval.append(1)

    new_base = (2**31 - 1)**(1/count)

    prev = 1
    for i in range(count- 2):
        next = max(prev+1, int(new_base**(i+2)))
        retval.append(next)
        prev = next

    return retval
```
I tried using the above on my problem via: `gbm = lgb.LGBMRanker(label_gain = get_label_gain(300))` and for the first time I got the ranker to do some training without errors.  The ndcg values improved from 0.70 to 0.85 during training so it appears to be doing something. 

Besides binning my fitness values into 30 different bins (or more if I pass a custom label_gain) it seems like I could also just divide my 300 result dataset into 10 queries of 30 elements and then to calculate each result's place in the 30 element query.   It seems like grouping results in place 1-30 together as one query and then 31-60 together would be the wrong way to do it.  If it was done that way lightgbm would (presumably) spend as much effort trying to put 1 and 2 in order as it does to put 31 and 32 in order.  Which is not what I want.  So maybe query one would have [1,11,21,31...] and query 2 would be [2,12,22,32,... ]  etc.  but then the ranker never gets to see item 1 vs item 2 which is presumably the most important thing for the ranker to learn. So that's probably not the best way to do it either.  

I'm sure most of my questions can be answered by doing experiments but feel free to chime in. 

Thanks for everyone's hard work on this library.  Sorry the post has grown this long but maybe a detailed description of my problems will help improve the documentation or help the next person.

**Comment by shiyu1994** [comment]
@rripken Thanks for the detailed description. Yes, LightGBM restricts the maximum number of relevance score to 31 because the default `label_gain` is in the exponential form `2 ** relevance`. The restriction is to avoid the overflow problem.

As posted by @rripken, one option is to use linearly growing label gain instead of exponentially growing, or to change the base of the exponential (smaller than 2) so that `base ** max_relevance_score` does not overflow.
